drivers/rtc/rtc_ll_stm32.c

```diff
@@ -0,0 +1,360 @@
+
+	/* The longest period we can match for universally is the
+	   duration of the shortest month */
+	if ((alarm_val - now) > (RTC_ALARM_DAY * 28)) {
+		return -ENOTSUP;
+	}
+
+	/* Convert seconds since 2000-01-01 00:00:00 to calendar datetime */
+	alarm_ts = alarm_val;
+	alarm_ts += EPOCH_OFFSET;
+	gmtime_r(&alarm_ts, &alarm_tm);
+
+	/* Apply ALARM_A */
+	rtc_alarm.AlarmTime.TimeFormat = LL_RTC_TIME_FORMAT_AM_OR_24;
+	rtc_alarm.AlarmTime.Hours = alarm_tm.tm_hour;
+	rtc_alarm.AlarmTime.Minutes = alarm_tm.tm_min;
+	rtc_alarm.AlarmTime.Seconds = alarm_tm.tm_sec;
+
+	rtc_alarm.AlarmMask = LL_RTC_ALMA_MASK_NONE;
+	rtc_alarm.AlarmDateWeekDaySel = LL_RTC_ALMA_DATEWEEKDAYSEL_DATE;
+	rtc_alarm.AlarmDateWeekDay = alarm_tm.tm_mday;
+
+	LL_RTC_DisableWriteProtection(RTC);
+	LL_RTC_ALMA_Disable(RTC);
+
+	if (LL_RTC_ALMA_Init(RTC, LL_RTC_FORMAT_BIN, &rtc_alarm) != SUCCESS) {
+		return -EIO;
+	}
+
+	LL_RTC_DisableWriteProtection(RTC);
+
+	LL_RTC_ALMA_Enable(RTC);
+	LL_RTC_ClearFlag_ALRA(RTC);
+	LL_RTC_EnableIT_ALRA(RTC);
+
+	LL_RTC_EnableWriteProtection(RTC);
+
+	/* RTC Alarm Interrupt Configuration: EXTI configuration */
+	LL_EXTI_EnableIT_0_31(LL_EXTI_LINE_18);
+	LL_EXTI_EnableRisingTrig_0_31(LL_EXTI_LINE_18);
+
+	irq_enable(CONFIG_RTC_0_IRQ);
+
+	return 0;
+}
+
+static int rtc_stm32_set_config(struct device *dev, struct rtc_config *cfg)
+{
+	int result = 0;
+	time_t init_ts = 0;
+	struct tm init_tm = { 0 };
+
+	LL_RTC_DateTypeDef rtc_date;
+	LL_RTC_TimeTypeDef rtc_time;
+
+	/* Convert seconds since 2000-01-01 00:00:00 to calendar datetime */
+	init_ts = cfg->init_val;
+	init_ts += EPOCH_OFFSET;
+
+	gmtime_r(&init_ts, &init_tm);
+
+	rtc_date.Year = init_tm.tm_year % 100;
+	rtc_date.Month = init_tm.tm_mon;
+	rtc_date.Day = init_tm.tm_mday;
+
+	rtc_time.TimeFormat = LL_RTC_TIME_FORMAT_AM_OR_24;
+	rtc_time.Hours = init_tm.tm_hour;
+	rtc_time.Minutes = init_tm.tm_min;
+	rtc_time.Seconds = init_tm.tm_sec;
+
+	k_sem_take(DEV_SEM(dev), K_FOREVER);
+
+	if (cfg->cb_fn != NULL) {
+		DEV_DATA(dev)->cb_fn = cfg->cb_fn;
+	}
+
+	if (LL_RTC_DATE_Init(RTC, LL_RTC_FORMAT_BIN, &rtc_date) != SUCCESS) {
+		result = -EIO;
+		goto fin;
+	}
+
+	if (LL_RTC_TIME_Init(RTC, LL_RTC_FORMAT_BIN, &rtc_time) != SUCCESS) {
+		result = -EIO;
+		goto fin;
+	}
+
+	if (cfg->alarm_enable) {
+		rtc_stm32_set_alarm(dev, cfg->alarm_val);
+	}
+
+fin:
+	k_sem_give(DEV_SEM(dev));
+
+	return result;
+}
+
+static u32_t rtc_stm32_get_pending_int(struct device *dev)
+{
+	return LL_RTC_IsActiveFlag_ALRA(RTC) != 0;
+}
+
+void rtc_stm32_isr(void *arg)
+{
+	struct device *const dev = (struct device *)arg;
+
+	if (LL_RTC_IsActiveFlag_ALRA(RTC) != 0) {
+
+		if (DEV_DATA(dev)->cb_fn != NULL) {
+			DEV_DATA(dev)->cb_fn(dev);
+		}
+
+		LL_RTC_ClearFlag_ALRA(RTC);
+		LL_RTC_DisableIT_ALRA(RTC);
+
+	}
+
+	LL_EXTI_ClearFlag_0_31(LL_EXTI_LINE_18);
+}
+
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP
+static int rtc_suspend_device(struct device *dev)
+{
+	rtc_stm32_disable(dev);
```

Comment:
You should not disable RTC when entering Sleep mode, it will halt the RTC.
Low power should actually be handlded by the PWR IP which should map OS power management requests to SoC power states.
RTC behavior (and for other IP it is the same) will be altered depending on PWR state that will be defined (CF "RTC and low-power modes" in reference manual.
This being said, I think you should remove power management from this driver.